### PR TITLE
refactor: Added intermediate object prior to query building 

### PR DESCRIPTION
**Pull request by calumbrodie**
Added intermediate object prior to query building  which contains the SelectQuery and Context (ResolvedSelectQuery), refactor SolariumSelectQueryBuilder to use this new construct.

Also various CS and docblock fixes

### CODE REVIEW

File: Builder/SolariumSelectQueryBuilder.php [hunk 1]
```diff
@@ -71,33 +64,35 @@ public function buildSolariumQueryFromGeneric(GenericSelectQuery $query)
         }
 
         //determine whether we are using the facet values for an underlying "base" (recorded) query
-        $shouldUseFacetValuesForRecordedQuery = (bool) $this->hasSearchContext()
-            && $query instanceof RecordableSelectQueryInterface
+        $shouldUseFacetValuesForRecordedQuery = (bool) $query instanceof RecordableSelectQueryInterface
```
Comment (calumbrodie) [inline Builder/SolariumSelectQueryBuilder.php]: I couldn't determine why it was important/relevant that there was an applied search context while working out this value. I thought it was safe to remove it (since the context is no longer available in this class - by design).

File: Builder/SolariumSelectQueryBuilder.php [hunk 2]
```diff
@@ -71,33 +64,35 @@ public function buildSolariumQueryFromGeneric(GenericSelectQuery $query)
...
+                    $this->lucenifyFilterQueries($filterQueries),
+                    $this->lucenifyFilterQueries($query->getRecord()->getFilterQueries())
+                )
+            );
         }
 
-        if ($this->hasSearchContext()) {
-            //add default filter queries from the search context
-            $filterQueries = array_merge($filterQueries, $this->getSearchContext()->getDefaultFilterQueries());
-        }
```
Comment (calumbrodie) [inline Builder/SolariumSelectQueryBuilder.php]: These are now merged within ResolvedSelectQuery

File: Builder/SolariumSelectQueryBuilder.php (hunk 1 above)
Comment (shieldo) [inline Builder/SolariumSelectQueryBuilder.php]: Shouldn't this check be exposed by the `ResolvedSelectQueryInterface`?

File: Builder/SolariumSelectQueryBuilder.php [hunk 3]
```diff
@@ -111,96 +106,94 @@ public function buildSolariumQueryFromGeneric(GenericSelectQuery $query)
...
             $facetNamesToExclude = $query->getFacetNamesToExclude();
-            if (!empty($facets)) {
-                foreach ($facets as $facet) {
-                    //if query indicates we should skip this facet, then skip it
-                    if (false !== array_search($facet->getSearchKey(), $facetNamesToExclude)) {
-                        continue;
-                    }
-                    $solariumFacets = [];
-                    //check whether to request missing facet values
-                    $checkMissingFacetValues = (method_exists($this->getSearchContext(), 'shouldRequestFacetValueForMissing')) ? $this->getSearchContext()->shouldRequestFacetValueForMissing() : null;
```
Comment (calumbrodie) [inline Builder/SolariumSelectQueryBuilder.php]: I added this method to the SearchContextInterface and repaired all implementations. Considered this line a bit sketchy :)

File: Context/SearchContextInterface.php [hunk 4]
```diff
@@ -90,4 +90,11 @@ public function getFacetSortOrderProvider();
      * @return \Markup\NeedleBundle\Intercept\InterceptorInterface
      **/
     public function getInterceptor();
+
```
Comment (shieldo) [inline Context/SearchContextInterface.php]: This seems out of scope for a "refactor"

**Comment by shieldo** [comment]
I'd like to see some unit tests for the new resolved query object.

File: Builder/SolariumSelectQueryBuilder.php (hunk 1 above)
Comment (calumbrodie) [inline Builder/SolariumSelectQueryBuilder.php]: Yeah - maybe I should add the whole check to the 'ResolvedSelectQueryInterface'? I still can't see why the context is relevant though...

File: Context/SearchContextInterface.php (hunk 4 above)
Comment (calumbrodie) [inline Context/SearchContextInterface.php]: Why? I explained why this was added.

File: Builder/SolariumSelectQueryBuilder.php (hunk 1 above)
Comment (shieldo) [inline Builder/SolariumSelectQueryBuilder.php]: `getRecord` is being exposed - I guess if this returns null we can consider this query not recorded?

File: Context/SearchContextInterface.php (hunk 4 above)
Comment (shieldo) [inline Context/SearchContextInterface.php]: It's needed by the purpose of this PR (creating an intermediate object)?

I don't disagree with its inclusion, it just seems like an irrelevant change to this PR.

**Comment by calumbrodie** [comment]
Sure will add some tests for that class now

File: Context/SearchContextInterface.php (hunk 4 above)
Comment (shieldo) [inline Context/SearchContextInterface.php]: (Especially as it's a breaking change - it'll make any class implementing this fatal unless it's updated.  This PR shouldn't change the behaviour, just make it more flexible for decorating the behaviour)

Comment (calumbrodie) [inline Context/SearchContextInterface.php]: Have removed - will open a separate PR

**Comment by calumbrodie** [comment]
@shieldo Added tests, would like to merge.

**Comment by calumbrodie** [comment]
@shieldo I think Travis CI is broken? Not sure if these are actually failing (and the logs are just broken on Travis), or if they are not failing

**Comment by calumbrodie** [comment]
@shieldo NVM - it's broken, will fix :/
